test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go

```diff
@@ -0,0 +1,508 @@
+	}
+
+	if err := pool.DeleteExtraMachines(ctx); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to delete overprovisioned or out of spec machines")
+	}
+
+	dockerMachinePool.Spec.ProviderIDList = []string{}
+	for _, instance := range dockerMachinePool.Status.Instances {
+		if instance.ProviderID != nil && instance.Ready {
+			dockerMachinePool.Spec.ProviderIDList = append(dockerMachinePool.Spec.ProviderIDList, *instance.ProviderID)
+		}
+	}
+
+	dockerMachinePool.Status.Replicas = int32(len(dockerMachinePool.Spec.ProviderIDList))
+
+	if dockerMachinePool.Spec.ProviderID == "" {
+		dockerMachinePool.Spec.ProviderID = getDockerMachinePoolProviderID(cluster.Name, dockerMachinePool.Name)
+	}
+
+	dockerMachinePool.Status.Ready = len(dockerMachinePool.Spec.ProviderIDList) == int(*machinePool.Spec.Replicas)
+	return ctrl.Result{}, nil
+}
+
+type nodePool struct {
+	client            client.Client
+	cluster           *clusterv1.Cluster
+	machinePool       *clusterv1exp.MachinePool
+	dockerMachinePool *infrav1exp.DockerMachinePool
+	machines          []*docker.Machine
+	logger            logr.Logger
+}
+
+func newNodePool(kClient client.Client, cluster *clusterv1.Cluster, mp *clusterv1exp.MachinePool, dmp *infrav1exp.DockerMachinePool, log logr.Logger) (*nodePool, error) {
+	np := &nodePool{
+		client:            kClient,
+		cluster:           cluster,
+		machinePool:       mp,
+		dockerMachinePool: dmp,
+		logger:            log.WithValues("node-pool", dmp.Name),
+	}
+
+	if err := np.Refresh(); err != nil {
+		return np, errors.Wrapf(err, "failed to refresh the node pool")
+	}
+	return np, nil
+}
+
+func (np *nodePool) Refresh() error {
+	machinePoolLabels := map[string]string{"dmp": np.dockerMachinePool.Name}
+	machines, err := docker.ListMachinesByCluster(np.cluster.Name, machinePoolLabels, np.logger)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list all machines in the cluster")
+	}
+
+	np.machines = machines
+	return nil
+}
+
+func (np *nodePool) MachinesMatchingSpec() []*docker.Machine {
+	var matchingMachines []*docker.Machine
+	for _, machine := range np.machines {
+		if !machine.IsControlPlane() && machine.ImageVersion() == *np.machinePool.Spec.Template.Spec.Version {
+			matchingMachines = append(matchingMachines, machine)
+		}
+	}
+
+	return matchingMachines
+}
+
+func (np *nodePool) MachinesNotMatchingSpec() []*docker.Machine {
+	var matchingMachines []*docker.Machine
+	for _, machine := range np.machines {
+		if !machine.IsControlPlane() && machine.ImageVersion() != *np.machinePool.Spec.Template.Spec.Version {
+			matchingMachines = append(matchingMachines, machine)
+		}
+	}
+
+	return matchingMachines
+}
+
+func (np *nodePool) AddNewMachine(ctx context.Context) error {
+	instanceName := util.RandomString(6)
+	externalMachine, err := docker.NewMachine(np.cluster.Name, instanceName, np.dockerMachinePool.Spec.Template.CustomImage, map[string]string{"dmp": np.dockerMachinePool.Name}, np.logger)
```

Comment:
what about making "dmp" more expressive, so if someone inspect the container it is clear the link with CAPD
e.g. `docker.cluster.x-k8s.io/machine-pool`